Mathlib/Tactic/NormNum/BigOperators.lean

```diff
@@ -0,0 +1,372 @@
+  ext x; simpa using complete x
+
+/-- Either show the expression `s : Q(Finset α)` is empty, or remove one element from it.
+
+Fails if we cannot determine which of the alternatives apply to the expression.
+-/
+partial def Finset.proveEmptyOrCons {α : Q(Type u)} :
+  (s : Q(Finset $α)) → MetaM (Finset.ProveEmptyOrConsResult s) :=
+fun s ↦
+match Expr.getAppFnArgs s with
+| (`EmptyCollection.emptyCollection, _) => pure (.empty (q(rfl) : Q($s = $s)))
+| (`Finset.cons, #[_, a, s', h]) => pure (.cons a s' h (q(rfl) : Q($s = $s)))
+| (`Finset.mk, #[_, (val : Q(Multiset $α)), (nd : Q(Multiset.Nodup $val))]) => do
+  match ← Multiset.proveZeroOrCons val with
+  | .zero pf => pure <| .empty (q($pf ▸ Finset.mk_zero) : Q(Finset.mk $val $nd = ∅))
+  | .cons a s' pf => do
+    let h : Q(Multiset.Nodup ($a ::ₘ $s')) := q($pf ▸ $nd)
+    let nd' : Q(Multiset.Nodup $s') := q((Multiset.nodup_cons.mp $h).2)
+    let h' : Q($a ∉ $s') := q((Multiset.nodup_cons.mp $h).1)
+    return (.cons a q(Finset.mk $s' $nd') h'
+      (q($pf ▸ Finset.mk_cons $h) : Q(Finset.mk $val $nd = Finset.cons $a ⟨$s', $nd'⟩ $h')))
+| (`Finset.range, #[(n : Q(ℕ))]) => do
+  match ← Nat.unifyZeroOrSucc n with
+  | .zero _pf => do
+    pure (.empty (q(Finset.range_zero) : Q(Finset.range 0 = ∅)))
+  | .succ n' _pf => pure <| ((Finset.ProveEmptyOrConsResult.cons
+      n'
+      (q(Finset.range $n'))
+      (q(@Finset.not_mem_range_self $n'))
+      (q(Finset.range_succ_eq_cons $n'))).uncheckedCast
+        (q(Finset.range (Nat.succ $n')) : Q(Finset ℕ))
+        s)
+| (`Finset.univ, #[_, instFT]) => do
+  match Expr.getAppFnArgs (← whnfI instFT) with
+  | (`Fintype.mk, #[_, (elems : Q(Finset $α)), (complete : Q(∀ x : $α, x ∈ $elems))]) => do
+    have _instFT : Q(Fintype $α) := instFT
+    let res ← Finset.proveEmptyOrCons elems
+    pure <| res.eq_trans (q(Finset.univ_eq_elems $elems $complete) : Q(Finset.univ = $elems))
+  | e => throwError "Finset.proveEmptyOrCons: could not determine elements of Fintype instance {e}"
+| (fn, args) =>
+  throwError "Finset.proveEmptyOrCons: unsupported finset expression {s} ({fn}, {args})"
+
+namespace NormNum
+
+/-- If `a = b` and we can evaluate `b`, then we can evaluate `a`. -/
+def Result.eq_trans {α : Q(Type u)} {a b : Q($α)} (eq : Q($a = $b)) : Result b → Result a
+| .isBool true proof =>
+  have a : Q(Prop) := a
+  have b : Q(Prop) := b
+  have eq : Q($a = $b) := eq
+  have proof : Q($b) := proof
+  Result.isTrue (x := a) q($eq ▸ $proof)
+| .isBool false proof =>
+  have a : Q(Prop) := a
+  have b : Q(Prop) := b
+  have eq : Q($a = $b) := eq
+  have proof : Q(¬ $b) := proof
+ Result.isFalse (x := a) q($eq ▸ $proof)
+| .isNat inst lit proof => Result.isNat inst lit q($eq ▸ $proof)
+| .isNegNat inst lit proof => Result.isNegNat inst lit q($eq ▸ $proof)
+| .isRat inst q n d proof => Result.isRat inst q n d q($eq ▸ $proof)
+
+protected lemma Finset.sum_empty {β α : Type _} [CommSemiring β] (f : α → β) :
+    IsNat (Finset.sum ∅ f) 0 :=
+  ⟨by simp⟩
```

Comment:
Does this work via `to_additive`?

Reply:
I get the following error:
```
application type mismatch
  Finset.sum
argument has type
  CommMonoid β
but function has type
  [inst : AddCommMonoid β] → Finset α → (α → β) → β
```
I think it's because the `CommSemiring` argument, that `CommSemiring.toCommMonoid` does not get `to_additive`ised to `CommSemiring.toAddCommMonoid`.